server/app/services/program/ProgramServiceImpl.java

```diff
@@ -705,6 +789,86 @@ public ProgramDefinition setBlockEligibilityDefinition(
     return updateProgramDefinitionWithBlockDefinition(programDefinition, blockDefinition);
   }
 
+  @Override
+  public ProgramDefinition setBlockEligibilityDefinition(
+      long programId, long blockDefinitionId, DynamicForm predicateForm)
+      throws ProgramNotFoundException, ProgramBlockDefinitionNotFoundException,
+          IllegalPredicateOrderingException {
+    ProgramDefinition programDefinition = getProgramDefinition(programId);
+
+    PredicateDefinition predicateDefinition = generatePredicateDefinition(predicateForm);
+    EligibilityDefinition eligibility =
+        EligibilityDefinition.builder()
+            .setPredicate(predicateDefinition)
+            .setPredicateFormat(SINGLE_LAYER_AND)
+            .build();
+
+    BlockDefinition blockDefinition =
+        programDefinition.getBlockDefinition(blockDefinitionId).toBuilder()
+            .setEligibilityDefinition(eligibility)
+            .build();
+
+    return updateProgramDefinitionWithBlockDefinition(programDefinition, blockDefinition);
+  }
+
+  private static final Pattern PREDICATE_VALUE_FORM_KEY_PATTERN =
+      Pattern.compile("^group-(\\d+)-question-(\\d+)-predicateValue$");
+
+  private PredicateDefinition generatePredicateDefinition(DynamicForm predicateForm) {
```

Comment:
this appears to be static

Reply:
With a few notable exceptions (build classes, very high-level utility functions), I prefer to introduce new dependencies with an object-oriented public interface (i.e. using instance methods), regardless of whether or not they have internal state. This reduces the cost of introducing state to them in the future since consumers of their public methods won't need to be updated.

Reply:
moved to utility method